src/condor_schedd.V6/qmgmt_startup_limits.cpp

```diff
@@ -0,0 +1,610 @@
+        double rate = capacity / window.count();
+        double elapsed = std::chrono::duration_cast<std::chrono::duration<double>>(now - last_refill).count();
+        tokens = std::min(capacity, tokens + rate * elapsed);
+        tokens = std::max(-burst, tokens); // do not accumulate deeper debt than burst
+        last_refill = now;
+    }
+
+    bool allow(std::chrono::steady_clock::time_point now, double cost) {
+        refill(now);
+        double need = std::max(0.0, cost);
+        if (max_burst_cost > 0) {
+            need = std::min(need, max_burst_cost);
+        }
+        double available = tokens + burst;
+        if (available + 1e-9 < need) { return false; }
+        tokens -= need;
+        tokens = std::max(-burst, tokens);
+        return true;
+    }
+
+    int capacityIn(std::chrono::duration<double> horizon, std::chrono::steady_clock::time_point now) {
+        refill(now);
+        double rate = (window.count() > 0) ? (capacity / window.count()) : 0.0;
+        double future = tokens + rate * std::max(0.0, horizon.count());
+        future = std::min(capacity, future);
+        future = std::max(-burst, future);
+        double available = std::min(capacity + burst, future + burst);
+        return (int)floor(available + 1e-6);
+    }
+};
+
+struct StartupLimit {
+    std::string uuid;
+    std::string tag;
+    std::string display_name;      // human-friendly name; falls back to uuid
+    std::string expr_source;       // rewritten expression string
+    std::unique_ptr<classad::ExprTree> expr;
+    std::string cost_expr_source;
+    std::unique_ptr<classad::ExprTree> cost_expr;
+    classad::References job_refs;
+    classad::References target_refs;
+    bool job_only{false};          // true if expression references only job attrs
+    RateLimit rate;
+    double burst{0};               // tokens allowed below zero
+    double max_burst_cost{0};
+    time_t expires_at{0};
+    long long jobs_skipped{0};
+    long long matches_ignored{0};
+    long long block_seconds{0};
+    long long matches_ignored_reported{0};
+    long long block_seconds_reported{0};
+    time_t last_report_time{0};
+    time_t last_ignored{0};
+    std::unordered_map<std::string, time_t> ignored_sources; // key=user|pool
+};
+
+static std::unordered_map<std::string, StartupLimit> StartupLimits;
+static int StartupLimitCleanupTid = -1;
+static int StartupLimitMaxExpiration = 0; // seconds
+static int StartupLimitBanWindow = 60;     // seconds after an ignored match
+static int StartupLimitLookahead = 60;     // seconds for throughput throttling
+
+static std::string make_source_key(const char *user, const char *pool) {
+    std::string key = user ? user : "";
+    key += "|";
+    if (pool) { key += pool; }
+    return key;
+}
+
+static std::string new_startup_uuid() {
+    unsigned long long now = (unsigned long long)time(nullptr);
+    unsigned long long rnd = ((unsigned long long)getpid() << 32) ^ (unsigned long long)random();
```

Comment:
```suggestion
    unsigned long long rnd = ((unsigned long long)getpid() << 32) ^ (unsigned long long)get_random_uint_insecure();
```
This is what we need for Windows.